src/main/scala/formal/FormalUtils.scala

```diff
@@ -0,0 +1,247 @@
+// See LICENSE for license details.
+package freechips.rocketchip.formal
+
+import Chisel._
+import chisel3.core.{VecInit}
+import chisel3.util.Cat
+import chisel3.internal.sourceinfo.{SourceInfo, SourceLine}
+
+sealed abstract class MonitorDirection(name: String) {
+  override def toString: String = name
+  def flip: MonitorDirection
+}
+object MonitorDirection {
+  // Also known as master, effectively contains assumes
+  object Driver  extends MonitorDirection("Driver") { override def flip: MonitorDirection = Receiver }
+
+  // Also known as slave, effectively contains asserts
+  object Receiver extends MonitorDirection("Receiver") { override def flip: MonitorDirection = Driver }
+
+  object Monitor extends MonitorDirection("Monitor") { override def flip: MonitorDirection = Monitor }
+
+  object Cover extends MonitorDirection("Cover") { override def flip: MonitorDirection = Cover }
+}
+
+sealed abstract class PropertyClass(name: String) {
+  override def toString: String = name
+}
+
+object PropertyClass {
+  object Default extends PropertyClass("Default")
+  // Properties which should be true by local construction of RTL (not test bench dependent)
+  object LocalRTL extends PropertyClass("LocalRTL")
+  // Properties that is expected to fail
+  object Failure extends PropertyClass("Fail")
+  // Properties that may fail because DUT or test is not complete
+  object TODO extends PropertyClass("TODO")
+  // Properties that may fail because DUT is not complete
+  object TODO_RTL extends PropertyClass("TODO_RTL")
+  // Disable a cover property and turn it into an assert (for unreachable covers)
+  object CoverDisableMonitor extends PropertyClass("CoverDisableMonitor")
+}
+
+object Property {
+  var prop_name_set = collection.mutable.Set[String]()
+  def reset_prop_name: Unit = prop_name_set = collection.mutable.Set[String]()
+
+  def apply(dir: MonitorDirection, cond: Bool, message: String,
+            prop_type: PropertyClass=PropertyClass.Default,
+            idx: String = "")(implicit sourceInfo: SourceInfo): Unit = {
+    val line_info = sourceInfo match {
+      case SourceLine(filename, line, col) => s"${filename}_L${line}_C${col}_I${idx}".replace('.', '_')
+      case _ => ""
+    }
+    val proposed_src = prop_type.toString + "_" + line_info
+
+//    var max_rand = 15
+//    val random_src = scala.util.Random
+//    while (!prop_name_set.add(proposed_src)) {
+//      proposed_src = src + "_" + "%x".format(random_src.nextInt(max_rand))
+//      max_rand = max_rand * 2
+//    }
```

Comment:
Can you delete this commented out code? (FYI using `scala.util.Random` means your tests wont be reproducible from a known seed).